---
title: "RNA-seq analysis of Ewing's sarcoma"
author: "Victor Casneuf - Helen Verbeke - Tika Stuyvaert - Josse Verboven"
date: "08-05-2023"
output:
  pdf_document: default
html_document:
  df_print: paged
---

```{r global_options, eval=TRUE, include=F, message=T, warning=T}
#include=FALSE: This specifies that the code chunk should not be included in the final rendered document.
#echo = TRUE: This specifies that the code itself should be included in the final rendered document. If echo is set to FALSE, the code will not be shown in the final output.
#message = TRUE: This specifies that any messages generated by the R code within the chunk will be displayed in the output document.
#warning=TRUE: This specifies that any warnings that occur during code execution should be displayed in the final rendered document.
knitr::opts_chunk$set(fig.align = 'center')
#This line sets the default alignment for figures in code chunks to be centered. 
#This means that any plots or images produced by R code chunks will be centered on the page by default.
knitr::opts_chunk$set(out.width = '85%')
#This line sets the default output width for code chunk results to be 85% of the width of the output area.
#This means that any printed output, tables or plots produced by R code chunks will be scaled to 85% of the width of the output area by default.
```


# 1. Introduction

*Group 17*
*Helen Verbeke, Tika Stuyvaert, Josse Verboven, Victor Casneuf*

This document is the second part of our genome analysis project, where we investigate differential expression in Ewing sarcoma. In the first part, we analyzed a microarray data set, whereas here, we will analyze an RNA-sequencing data set. For this sequencing part we found a data set comparingEwing  sarcoma cell lines to mesenchymal stem cell lines as control samples.

## Loading necessary packages

Here we load the necessary packages for completing the differential expression analysis in R.

```{r, name = "Packages", message=FALSE, warning=FALSE}
# RNA-seq analysis
library("tximport")
library("biomaRt")
library("edgeR")
# Useful
library("dplyr")
```

# 2. Pre-processing in mobaXterm

We retrieved our data from the ArrayExpress repository:
https://www.ebi.ac.uk/biostudies/arrayexpress/studies/E-GEOD-73610?query=E-GEOD-73610

This data set contains only a few samples: two ES and three MSC samples. The sequencing was done with the Illumina HiSeq2500 platform and paired-end sequencing was performed. We downloaded the **raw data** as fastQ files to start out in MobaXterm for the pre-processing of the RNA-seq data. Because these pre-processing steps have to be performed on tens of millions of reads and thus require a big amount of computing power and memory, we got access to HPC infrastructure through a virtual organisation (VO) on Vlaams Supercomputer Centrum (VSC). After establishing a sftp connection in mobaXterm, we could upload our files and start running job scripts. 

## Quality control on fastQ files

We will be using the kallisto aligning algorithm, since we are just interested in counts and this algorithm only requires little computing power and memory (in comparison to STAR alignment for example). In mobaXterm, we downloaded the human reference genome and a GTF-file, to annotate the data, from Ensembl and performed quality control with the FastQC toolkit. These FastQC reports show high base calling quality, with Phred scores consistently above 34. The second ES sample (SRR2541171) has a minor spatial artifact but apart from that there are no great differences between the samples. Removing the spatial artifact would probably not be beneficial anyway since we already do not have a lot of data. The reads do not contain their adapter anymore and the peaks of the GC-content distribution seems to be around the same value in the different samples. The one thing that is striking is that some of the samples have different read lengths. This is odd, but should not be problematic. Overall, we can see that the raw data is of good quality.

## Pseudo-alignment

After performing read trimming with the Trimmomatic tool and pseudo-alignment with kallisto, we can evaluate the mapping statistics and re-evaluate the read quality, this time with MultiQC. We can see in the MultiQC report that the percentage of unique reads over the different samples lies between 60% and 70%, which is somewhat on the low side but not problematic. Now we can also clearly see that the average GC-content lies between 45% and 48% with the GC-content distribution plots differing very slightly between the MSC and ES samples. Note that SRR2541168 and SRR2541169 are the MSC samples and SRR2541170, SRR2541171 and SRR2541172 the ES samples.


```{r, name = "MultiQC"}
htmltools::includeHTML("./multiqc_report.html")

```

# 3. Data import into R

## Gene annotation

The kallisto files contain pseudo-counts. To summarize this data, we must continue in R. We want to summarize the data on the gene level rather than transcript level so we obtain the annotation data from the Ensembl database using the biomaRt package. We then read in the pseudo-count data using the tximport package and link the transcript IDs to their corresponding gene IDs. The pseudo-count table shows the number of reads mapped to each gene and can be used for downstream analysis such as differential gene expression analysis.

```{r, name = "Gene annotation"}
# Get annotation data
ensembl <- useMart("ensembl", dataset = "hsapiens_gene_ensembl", 
                   host="https://www.ensembl.org")

# Available attributes
atr <- listAttributes(ensembl)

data <- getBM(attributes = c('ensembl_gene_id', 'ensembl_transcript_id', 
                             'external_gene_name'), 
              mart = ensembl)

tx2geneGtf <- dplyr::select(data, ensembl_transcript_id, ensembl_gene_id)
tx2geneGtf <- dplyr::rename(tx2geneGtf, TXNAME = ensembl_transcript_id)
tx2geneGtf <- dplyr::rename(tx2geneGtf, GENEID = ensembl_gene_id)

head(tx2geneGtf)

# Get file locations
files <- list.files("./multiqc/kallisto_quant/")
files <- files[grep("abundance.tsv", files)]
samples <- unlist(strsplit(files, "-"))
files <- paste(rep("./multiqc/kallisto_quant/", length(files)), files, sep="")
names(files) <- samples

# Load RNAseq data
txi <- tximport(files, type = "kallisto", tx2gene = tx2geneGtf)

head(txi$counts)
dim(txi$counts)
```

## Sample annotation

Here we annotate the samples. This table shows again which SRR code belongs to which sample. This information is important for downstream analysis, as it allows us to compare and contrast gene expression patterns across the correct sample groups. Additionally, the table shows which exact cell lines were used for the samples.

```{r, name = "Sample annotation"}
sdrf_rnaseq <- read.delim("./E-GEOD-73610/E-GEOD-73610.sdrf.txt")
print(sdrf_rnaseq[,c("Comment..ENA_RUN.","Characteristics..organism.","Characteristics..cell.line.",
                     "Comment..Sample_source_name.")])
```

# 4. Statistical analysis with EdgeR

## Data preparation

Since we are working with pseudo-counts and not counts, we have to adapt the data to be compatible with the EdgeR package. Here we revert the normalization of the counts according to their read length. 

```{r, name = "Normalization"}
cts <- txi$counts
normMat <- txi$length

print(colnames(cts))

# Obtaining per-observation scaling factors for length, adjusted to avoid 
# changing the magnitude of the counts
normMat <- normMat/exp(rowMeans(log(normMat)))
normCts <- cts/normMat

# Computing effective library sizes from scales counts, to account for
# composition biases between samples
eff.lib <- calcNormFactors(normCts)*colSums(normCts)
# Calculate normalization factors for RNA-seq count data. 
# These factors are used to adjust for differences in library size and sequencing depth across samples, so that the counts can be compared between samples on a more equal footing.
# calcNormFacors produces a vector of normalization factors, with one factor for each sample in the matrix.

# Combining effective library sizes with the length factors, and calculating
# offsets for a log-link GLM
normMat <- sweep(normMat, 2, eff.lib, "*") 
normMat <- log(normMat)
```
Then we summarize the data by creating a DGElist object summarizing count data, gene IDâ€™s and sample annotation. We will also filter out the low-expression genes so that a matrix of CPM values can be created. 

```{r, name = "DGEList object"}
# Creating a DGEList object which can be used in EdgeR
y <- DGEList(cts)
y$samples$group <- as.factor(c("MSC", "MSC", "EWS", "EWS", "EWS"))

y <- scaleOffset(y, normMat)

# Filtering out low-level expression genes
keep <- filterByExpr(y)
y <- y[keep, ]
# y is now ready for estimate dispersion functions (edgeR user's guide)

dim(y)
```
After filtering out the low-level expression genes, we are left with 19.137 genes. These genes will be subjected to statistical analysis.

A matrix of CPM (counts per million) values is now created. CPM values are a normalized measure of gene expression that allows for comparison of expression levels between samples, accounting for differences in sequencing depth.

```{r, name = "CPM matrix"}
cpms <- edgeR::cpm(y, offset = y$offset, log = FALSE)  
```

Additionally, the design matrix is created, which specifies the relationship between the response variable and the predictor variables.

```{r, name = "Design matrix"}
y$samples$group <- as.factor(y$samples$group)
design_y <- model.matrix(~0+group, data = y$samples)
colnames(design_y) <- levels(y$samples$group)
design_y
```
## Dispersion estimation 

Going further, we can create a BCV (biological coefficient of variation) plot which shows the biological coefficient of variation in function of the average log CPMs. A higher BCV indicates a greater degree of variability in expression levels. The plot can be used to evaluate the quality of the data and to identify potential outliers or other issues that may need to be addressed in downstream analysis.

```{r, name = "BCV plot"}
# Estimate dispersions
y <- estimateDisp(y, design = design_y)
names(y)

# BCV plot
plotBCV(y)
```
Our BCV plot looks quite normal, with a common trend line that is relatively flat, indicating that there is little variation between samples. The trend line is also relatively stable across the range of average log CPM values, suggesting that the BCV is not strongly dependent on gene abundance. We notice that the trend line has a small variation which can indicate some variability but we do not believe this is a cause for any worry.

We use a GLM to model our expression data. 

```{r, name = "GLM"}
# Fit model
fit <- glmQLFit(y, design_y)
names(fit)

# Plot quasi-likelihood dispersions
plotQLDisp(fit)
```
This plot suggests that there is barely any dependance of the variation of the expression on the average expression values. Even though EdgeR takes this trended variance into account, this shows good quality.

## Statistical analysis

Finally, we perform a quasi-likelihood F-test to calculate the differential expression between ES and MSC on a 5% significance level. The topTags() function is used to correct for multiple testing using a false discovery rate (FDR) adjustment with a threshold of 0.05. 

```{r, name = "Statistical testing"}
# Contrast matrix
cont.matrix <- makeContrasts("EWS - MSC", levels=design_y)
cont.matrix

# Quasi-likelihood F-tests
qlf <- glmQLFTest(fit, contrast=cont.matrix)
sum(qlf$table$PValue < 0.05)

# Correcting for multiple testing (FDR < 0.05)
top.qlf <- topTags(qlf, n = Inf) # default sort by p-value
sum(top.qlf$table$FDR < 0.05)
```
The number of significant genes before multiple testing correction is 4.573. After multiple testing correction, 1.331 genes remain. Remember that we tested a total of 19.137 genes.

# 5. Visualisation

We perform some general plots for visualization purposes to summarize and visualize the results. We start with the MA plot.

```{r, name = "MA plot"}
# MA plot
with(qlf$table, plot(logCPM, logFC, pch=16, cex=0.2))
with(qlf$table, points(logCPM[p.adjust(PValue, method = "fdr")<0.05], 
                       logFC[p.adjust(PValue, method = "fdr")<0.05],
                       pch=16, col="red", cex=0.2))
```
Most genes cluster around an M value of zero, which indicates that for most genes, there is no differential expression between the sarcoma cell lines and the control cell lines. Though, we do see some significantly over- and underexpressed genes visualized as red dots in the MA plot. The two quasi-symmetrical lines underneath and above the gene cluster catch the attention and most likely indicate some kind of bias. It is possible that this is due to the spatial artifact identified in the FastQC report of the second ES sample. These genes do not seem to be the major proportion of the significant genes (red dots) and significant genes would have to be further tested on their biological role anyway so this is not a huge problem. 

Next, we utilize a volcano plot to have another way of visualizing the differential expression.

```{r, name = "Volcano plot"}
# Volcano plot
with(qlf$table, plot(logFC, -log(PValue), pch=16, cex=0.2))
with(qlf$table, points(logFC[p.adjust(PValue, method = "fdr")<0.05], 
                       -log(PValue)[p.adjust(PValue, method = "fdr")<0.05],
                       pch=16, col="red", cex=0.2))
```
The volcano plot looks decent. It has the distinct V-shape with some of the most significant points (red dots) also located the most on the outside, indicating that the values with the lowest p-values have the highest log fold change. 

We also make a histogram of the p-values, which will give an idea of the proportion of the statistically relevant in comparison to the non-relevant genes. 

```{r, name = "Histogram"}
# Distribution of the p-values (without correcting for multiple testing)
hist(qlf$table$PValue)
```
We see that there are a relatively big number of genes with low p-values (righ-skewed distribution). This is desired as a normal distribution would not result in many biologically relevant results.

Finally, we generate a box plot that compares the expression of the gene with the highest differential expression between the two groups (Ewing sarcoma and MSC). We also generate a box plot that shows the log fold change of the differentially expressed genes (FDR > 0.05) to get an idea of what value to select for the upcoming GSA.

```{r, name = "Box plots"}
# Comparison between the case and control group
boxplot(cpms[which(rownames(cpms)==rownames(top.qlf$table)[1]),]~y$samples$group,
        ylab="CPM",xlab="Group")

# Log fold change boxplot
boxplot(top.qlf$table$logFC[which(top.qlf$table$FDR < 0.05)], xlab="Significant genes (FDR < 0.05)", ylab="logFC")
```
The first boxplot shows that the gene with the highest differential expression between the EWS and MSC groups, has higher expression levels in the MSC group, which means this gene is probably downregulated in the sarcoma cells. The second boxplot shows the logFC distributions. This gives an idea on what values to pick for GSA.

## 6. Gene set analysis

For GSA, we will use the Webgestalt tool. We have to separate an object containing all gene names and an object containing a subset of the differentially expressed genes. We select an FDR of 0.05 and a log fold change 7. The log fold change is very arbitrary and was chosen based on how many genes the subset eventually consisted of. We aimed for a subset between 100 and 500 genes. Furthermore, we used an absolute logFC since there is not definite indication for systematic over- or underexpression. With these values, we get a collection of 398 genes.

```{r, name = "Gene selection"}
# Add column with IDs to EdgeR results
res <- top.qlf$table
res$ID <- row.names(res)
row.names(res) <- c(1:dim(res)[1])
res <- res[, c(6,1:5)]

# Select gene IDs for the background (i.e. all expressed genes)
gene_ids_background <- res$ID

# Select gene IDs for the set of interest (general disruption)
gene_ids_sign <- res$ID[(res$FDR < 0.05) & (abs(res$logFC) > 7)] 

gene_ids_sign <- gene_ids_sign[!is.na(gene_ids_sign)]

gene_ids_sign <- sort(as.character(gene_ids_sign))

length(gene_ids_sign)

```

The (Ensembl) gene IDs are written to .txt files which can be imported into the Webgestalt tool (select for Ensembl gene IDs) for GSA.

```{r, name = "txt files"}
# Write IDs to file
write.table(gene_ids_background, file="gene_ids_background.txt", col.names=F, quote=F, row.names=F)

write.table(gene_ids_sign, file="gene_ids_sign.txt",
col.names=F, quote=F, row.names=F)
```

https://www.webgestalt.org/results/1683407229/# 

The results of the GSA should be interpreted with care, since we have low power and many of the identified pathways have an FDR above 0.05. They serve as a proof-of-concept and indication for further research, rather than statistically robust evidence of the effect on certain pathways. That being said, the results seem to make sense, with obvious pathways such as transcriptional misregulation and proteoglycans in cancer, and inflammatory mediator regulation of T cells. Other pathways also point in an interesting direction: malaria is for example suggested to be linked to cancer. Additionally, the neuroactive ligand receptor interaction pathway and calcium signaling pathway have both been linked to cancer as well. 


## 7. Conclusion

Starting from the raw reads, the data showed decent quality. A spatial abnormality was observed, which could have lead to a bias later in the analysis, but it is not clear what the impact was. The main problem is the lack of samples which limits the ability to conclude significance, but the MA plot has shown that there is reason to believe that a decent portion of the differentially expressed genes actually has biological relevance. Because of these limitations, it is hard to pick out single genes, but together with GSA the overall difference in gene expression is clear. 
