---
title: "Microarray analysis of Ewing Sarcoma"
author: "Victor Casneuf - Helen Verbeke - Tika Stuyvaert - Josse Verboven"
date: "08-05-2023"
output:
  pdf_document: default
html_document:
  df_print: paged
—--

```{r global_options, eval=TRUE, include=FALSE, message=TRUE, warning=TRUE}
#include=FALSE: This specifies that the code chunk should not be included in the final rendered document.
#echo = TRUE: This specifies that the code itself should be included in the final rendered document. If echo is set to FALSE, the code will not be shown in the final output.
#message = TRUE: This specifies that any messages generated by the R code within the chunk will be displayed in the output document.
#warning=TRUE: This specifies that any warnings that occur during code execution should be displayed in the final rendered document.
knitr::opts_chunk$set(fig.align = 'center')
#This line sets the default alignment for figures in code chunks to be centered. 
#This means that any plots or images produced by R code chunks will be centered on the page by default.
knitr::opts_chunk$set(out.width = '85%')
#This line sets the default output width for code chunk results to be 85% of the width of the output area.
#This means that any printed output, tables or plots produced by R code chunks will be scaled to 85% of the width of the output area by default.
```

# 1. Introduction

*Group 17*
*Helen Verbeke, Tika Stuyvaert, Josse Verboven, Victor Casneuf*

Ewing sarcoma (ES) is a rare and aggressive type of cancer, discovered by James Ewing. It is mostly located in bones and primarily affects children and young adults. Although treatments have been drastically improved, ES remains difficult to treat. The disease is characterized by chromosomal translocations resulting in fused genes. The main mutation is the fusion of the genes EWSR1 and FLI1 and together with other fusion genes, the disease dysregulates various pathways. It is believed that ES originates from mesenchymal stem cells (MSC). 
 
For this project, we used two data sets to dive deeper into the gene expression patterns in Ewing sarcoma, in comparison to mesenchymal stem cells. One is from a microarray assay, the other one from RNA sequencing. We explored these data sets in an attempt to identify genes and pathways that are differentially expressed in relation to ES. The focus of this project is mainly on using the right methods for data exploration and testing, and also deriving the right conclusions. 


## Loading necessary packages

In order to carry out our study, we will load the requisite packages. As the Affymetrix Human Exon 1.0 ST Array was utilized for this particular experiment, it is necessary to install and load the oligo package.

```{r, name = "packages", message = FALSE, warning=FALSE}
# Microarray analysis
library("affy")
library("affyPLM")
library("arrayQualityMetrics") #Quality control
library("ALLMLL") #Data resource
library("oligo")
library("GEOquery")
library("limma")

# Useful
library("tidyverse") #dplyr and ggplot
library("reshape2")
library("gridExtra")
library("RColorBrewer")
library("genefilter")
```

# 2. Data exploration

We found our data on the gene expression omnibus (GEO) repository: https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE68776&fbclid=IwAR2bxec0TAOYh355P6uIU5bBIZd1RAzGAv5K60J6s2PzN8uZNS7ZwrINCww

First of all we will be loading in the data. We only used 38 of the original 74 samples of the raw data. Our data includes 32 archived tumor biopsy specimens obtained from patients with localized Ewing sarcoma and 6 controls. The controls are MSC samples in two triplicates.


```{r, name = "load in data", message = FALSE}
load("./GSE68776/Ewing.Rdata")
load("./GSE68776/gse.Rdata")
```

After importing the microarray data, we extract the specific portion of the phenotype data that is related to our study. In this case, we are solely interested in the phenotype data related to Ewing sarcomas and the controls. Due to the large size of the data, we have chosen to reduce our sample size from 38 to 21. As a result, we will be focusing on the EWS samples numbered 18 through 32, along with the six control samples.

```{r, name = "Ewing exploration", warning=FALSE}
# Exploration of the array express
class(Ewing)
car::some(exprs(Ewing), n = 10, cols = 6)

# Changing Sample names
sampleNames(Ewing) <- Ewing$title
sampleNames(Ewing)[33:38] <- c("control1", "control2", "control3", "control4", "control5", "control6")
Ewing <- Ewing[,!(sampleNames(Ewing) %in% 
                       c("EWS#1", "EWS#2", "EWS#3", "EWS#4", "EWS#5", "EWS#6", "EWS#7", "EWS#8", "EWS#9", "EWS#10", "EWS#11", "EWS#12", "EWS#13","EWS#14", "EWS#15", "EWS#16","EWS#17"))] 
```

```{r, name = "pheno"}
pheno <- pData(gse$GSE68776_series_matrix.txt.gz)
pheno <- pheno[c(18:35, 39:41),] %>%
  dplyr::select(c(title, source_name_ch1, characteristics_ch1))
# Remove old data
rm(gse)
# The microarrays and their responsible phenotype + tissue of origin
head(pheno)
```

# 3. Gene annotation

To report which genes are differentially expressed, gene annotation is performed. 

```{r, name = "annotation", message = FALSE}
annotation <- read.table("./GSE68776/GPL5175-3188.txt", sep = "\t", header = FALSE,
                         skip = 11, quote =, fill = T)
dim(annotation)
annotation[,2] <- gsub(",.+$","",annotation[,2])
colnames(annotation)[1] <- "PROBEID"
colnames(annotation)[2] <- "SYMBOL"
annotation$PROBEID <- as.character(annotation$PROBEID)
```

# 4. Quality control

Performing quality control is a crucial pre-processing step in conducting a gene expression study, as it ensures that experimental artifacts do not interfere with the statistical analysis. To achieve this, we will examine the microarray image plots, data distribution, and MA plots.

## 4.1 Microarray images

To begin the quality assurance, we examine the image plots of the data at the probe level to identify any significant spatial anomalies present in the individual arrays. The first three image plots represent the first 3 EWS samples and the bottom image plots are from the first three controls.

```{r, name = “images”}
par(mfrow = c(1,3))
for(i in 1:3){
  image(Ewing[,i])
}
par(mfrow = c(2,3))
for(i in 16:21){
  image(Ewing[,i])
}
```

### Conclusion images

Upon initial inspection, no visible anomalies are detected in the image plots. This implies that the expression data for the first three arrays and all six controls exhibit good quality.

## 4.2 Array quality metrics report of raw data

Since we have loaded in our data using the oligo package instead of affy, we cannot utilize the Affybatch class for further phenotypic information of the microarray. To perform further microarray quality control we use the arrayQualityMetrics function.

```{r, name = "ArrayQualityMetrics of Raw", warning=FALSE}
arrayQualityMetrics(expressionset = Ewing,
  intgroup = c('title', 'source_name_ch1'),
  outdir = "Quality report Ewing raw data",
  reporttitle = "arrayQualityMetrics report for the raw Ewing data",
  do.logtransform = TRUE, force = TRUE)
```

### Conclusion of array quality metrics report of raw data

It is always a trade-off between losing information and removing excessive variation.The report shows us that array 1 (sample EWS#18) is an obvious outlier which we think should best be removed. Removing array 12 (sample EWS#29) would also be a smart decision, which is made obvious by the box plots graph and the MA plot. So we remove these in the following chunk. The rest of the microarray analysis will be done without these and will thus conclude 19 samples.

```{r, name = "QC_Ewing"}
QC_Ewing <- Ewing[,!(sampleNames(Ewing) %in% 
                       c("EWS#18", "EWS#29"))] 
rm(Ewing)
```

## 4.3 Data distribution

Generating box plots and density plots of the probe-level data is a valuable next step. Our goal is to identify any arrays that may be defective, which we achieve by analyzing the box plots for any anomalies. We also examine the probe intensity behavior for any differences between the arrays. 

### Box plots

We will create **box plots** representing the perfect match (pm) intensity values for all arrays. This is done to detect the presence of potential detective arrays. Ideally, there should be a substantial overlap between the box plots with relatively few outliers.

```{r, name = "Box plot Raw"}
d.pm <- melt(oligo::pm(QC_Ewing), varnames=c('ProbeID', 'Array')) %>%
  mutate(Intensity = log2(value)) %>%
  mutate(Isolate = substr(Array, 1, 1)) 
p1.box <- ggplot(d.pm, aes(x = Array, y = Intensity)) +
  geom_boxplot(aes(fill = Array)) +
  labs(title = "Box Plots of sample Probe PM Level Data (not normalized)")+
  theme(legend.position = "none",
        axis.text.x = element_text(angle = 45),
        plot.title = element_text(size = 7))
p1.box
```
### Density plots

The next step involves generating and presenting the **density plots** of the pm intensity values. We look for any densities that are considerably distant from the others or that display distinctive shapes, such as bimodalities with two peaks.

```{r, name = "Density plot raw"}
p1.density <- ggplot(d.pm, aes(Intensity, group = Array)) + 
  stat_density(geom = 'line',
               position = 'identity',
               aes(colour=Isolate))+
  labs(x = "Log2 intensities",y ="Density",
       title = "Density plots pm probe level data for Microarrays")
p2.density <- ggplot(d.pm, aes(Intensity, group = Array)) + 
  stat_density(geom = 'line',
               position = 'identity',
               aes(colour=Array))+
  labs(x = "Log2 intensities",y ="Density",
       title = "Density plots pm probe level data for Microarrays") 
gridExtra::grid.arrange(p1.density, p2.density, nrow = 2)
```

### Conclusion of data distribution

A significant overlap between the box plots can be seen, although there are a few outliers. Sample 28 and 30 stand out a bit from the other box plots, as well as the 6 controls. This could mean a difference in probe intensity behavior between the arrays. This makes it hard to compare between samples. In the density plots, we observe some distant densities and different shapes maily when you look at the samples vs controls. It is remarkable that all the controls stand out again from the other densities. This difference may be attributed to the controls being obtained from different cells compared to the samples. Normalization is needed.


## 4.4 MA plots

The MA plot is a graphical representation that provides an overview of how the values compare to an assumed average of all the samples. It presents the log-fold change (M) in gene expression levels as a function of the log-intensity averages (A) for a specified array and the artificial array. In an ideal scenario, the majority of genes or probes should have an M-value of zero, which would indicate no differential expression between the compared conditions. However, for the differentially expressed genes we aim to identify, an absolute logFC higher than zero is expected. In fact, a logFC greater than one would be preferred for genes that are biologically relevant.

By plotting the MA plot, we can quickly determine if most genes indeed have an M-value of zero. If not, it could suggest that normalization is necessary to address any technical variations or biases that may be affecting the analysis.

```{r, name = "MA plot Raw", warning=FALSE}
expression <- exprs(QC_Ewing)
par(oma = c(1,1,3,1))
par(mfrow = c(2,3))
MAplot(expression, cex = .5)
mtext('M', side = 2, outer = TRUE)
mtext('A', side = 1, outer = TRUE)
mtext('Six MA-Plots vs. the Median Mock Array', side = 3, outer = TRUE)
```

### Conclusion MA plots

Ideally, there is close proximity between the horizontal 0 axis and the red line (loess) in the analysis. However, we have observed that this is not the case for certain samples: namely    EWS#22 and all the control samples deviate from this ideal. 
Furthermore, the distribution mass is not concentrated along the M = 0 axis, which implies that normalization may be required to account for technical variations and biases that could affect the interpretation of the biological significance of gene expression differences between the conditions being compared.

# 5. Microarray pre-processing

## 5.1 Background correction, normalization and data summarization

After conducting an initial quality control check on our microarray data, we observed some artifacts. To address this, we will carry out background correction, normalization, and summarization as crucial pre-processing steps for microarray data. Our goal is to ensure the accuracy and reliability of our microarray data and help avoid incorrect conclusions or false discoveries.

**Background correction** is a crucial initial step in processing microarray data, which aims to normalize the data by accounting for the intensity of the background signal surrounding each sample. This process helps to correct for variations in probe affinities, cross-hybridisation, and non-specific binding, which can otherwise skew the results of downstream analyses.

**Quantile normalization** ensures that the data are on a comparable scale and that differences between samples can be attributed to biological variation rather than technical artifacts. 

Last, a **data summarization** step (via median-polish) is needed to summarize multiple probe values in a single value per probeset. 

We will use the robust multi-array average function (rma) from the oligo package that performs these several preprocessing tasks at once.

```{r, name = "rma"}
ewing_rma <- oligo::rma(QC_Ewing)
save(ewing_rma, file= "./GSE68776/ewing_rma.Rdata")
load("./GSE68776/ewing_rma.Rdata")
```

## 5.2 Evaluation of pre-processing

After the background correction, normalization and data summarization of our data we repeat the quality control, to assess whether these preprocessing steps achieved its desired outcome. So next we'll review the **boxplots**, **density plots**, **MA-plots** and the **array quality metrics report** of the normalized data.

### Boxplots
```{r, name = "Boxplot rma"}
d.pm.norm <- melt(exprs(ewing_rma), varnames=c('ProbeID', 'Array')) %>%
  mutate(Intensity = log2(value)) %>%
  mutate(Isolate = substr(Array, 1, 1))
p2.box <- ggplot(d.pm.norm, aes(x = Array, y = Intensity)) + 
  geom_boxplot(aes(fill = Array)) +
  labs(title = "Box Plots of sample PM Level Data (normalized)")+
  theme(legend.position = "none",
        axis.text.x = element_text(angle = 45),
        plot.title = element_text(size = 7))
p2.box
```

### Density plots

```{r, name = "Density plot rma"}
p1.dens.norm <- ggplot(d.pm.norm, aes(Intensity, group = Array)) +
  stat_density(geom = "line",
               position = "identity",
               aes(colour=Isolate)) +
  labs(x="Log2 intensities",y="Density",
       title="Density plots normalized data")
p2.dens.norm <- ggplot(d.pm.norm, aes(Intensity, group = Array)) +
  stat_density(geom = "line",
               position = "identity",
               aes(colour=Array)) +
  labs(x="Log2 intensities",y="Density",
       title="Density plots normalized data (color per Array") +
  theme(legend.position="none")
gridExtra::grid.arrange(p1.dens.norm, p2.dens.norm, nrow = 2)
```

### MA plots

```{r, name = "MA-plots rma"} 
expression_rma <- exprs(ewing_rma)
par(oma = c(1,1,3,1))
par(mfrow = c(2,3))
MAplot(expression_rma, cex = .5)
mtext('M', side = 2, outer = TRUE)
mtext('A', side = 1, outer = TRUE)
mtext('Six MA-Plots vs. the Median Mock Array', side = 3, outer = TRUE) 
```

### Array quality metrics report of normalized data

```{r, ArrayQualityMetrics_of_rma, eval=TRUE, warning=FALSE}
arrayQualityMetrics(expressionset = ewing_rma,
  intgroup = c('title', 'source_name_ch1'),
  outdir = "Quality report Ewing normalized data",
  reporttitle = "arrayQualityMetrics report for the normalized Ewing data",
  do.logtransform = TRUE, force = TRUE)
```

### Conclusion of report from normalized data

Based on the obtained results after normalization, it is clear that the correction has been effective. The box plots and density profiles are similar now. The controls are no longer outliers. The MA plots also look better after the correction. Therefore, we will proceed to use the corrected, normalized, and summarized data for our analysis.


# 6. Differential expression analysis

In this study our objective is to identify consistent transcriptional changes that define Ewing Sarcoma cells, using a dataset that contains 32 tumor samples with localized Ewing sarcoma and 6 controls. So, we will look for differentially expressed genes by comparing the tumor samples with the controls. This is done with the limma package, which uses a linear model.

```{r, name = "DE"}
ewing_rma$source_name_ch1[1:13] <- "tumor"
ewing_rma$source_name_ch1[14:19] <- "MSC"
ewing_rma$source_name_ch1 <- factor(ewing_rma$source_name_ch1)
# Create the design matrix without an intercept
design <- model.matrix(~0 + source_name_ch1, data = ewing_rma)
# Create contrast matrix, where the control group is the reference group
contr.matrix <-makeContrasts(EWSvscontrol= source_name_ch1tumor - source_name_ch1MSC,
                            levels=colnames(design))

# Fit the linear model
# The contrast matrix is used to expand this linear model fit and find the DE genes 
# using Empirical Bayesian statistics. 
# The DE genes are found using a moderated F-test. 
fit <- lmFit(ewing_rma,design)
fit <- contrasts.fit(fit, contrasts = contr.matrix)
efit <- eBayes(fit)


list <- as.data.frame(rownames(exprs(ewing_rma)))
colnames(list)[1] <- "PROBEID" 
new <- left_join(list, annotation, by ="PROBEID")
Gene_names <- new$SYMBOL

res_moderated_t <- topTable(efit, sort.by = 'p', number = Inf)
# p-value less than 0.05, which is a common threshold for determining statistical significance in gene expression analysis. This means that only genes with a significant difference in expression between the two groups of samples are retained for further analysis.
sum(res_moderated_t$P.Value < 0.05)

# Get only the significant ones without and remove any duplicates from the subsetted data frame 
sign_genes <- unique(res_moderated_t[res_moderated_t$P.Value < 0.05,])
datanames <- rownames(sign_genes)
datanames <- as.integer(datanames)
ID_genes <- cbind(sign_genes,datanames)
colnames(ID_genes)[7] <- "PROBEID" 
ID_genes$PROBEID <- as.character(ID_genes$PROBEID)
ID_genes <- as.data.frame(ID_genes)
Sign_Genes_ID <- left_join(ID_genes,annotation, by = "PROBEID")
colnames(Sign_Genes_ID)[8] <- "GeneID"
Sign_Genes_ID <- dplyr :: select(Sign_Genes_ID, -c(V3, V5, V6, V7, V8, V9, V10, V11, V12))
Sign_Genes_ID <- Sign_Genes_ID[,c(8,7,1,2,3,4,5,6,7,9)]
head(res_moderated_t, 10)
head(Sign_Genes_ID, 10)
```
With the function decideTests, we can see how many up- and downregulated genes were detected. 

## 6.1 Multiple testing

When you perform multiple statistical tests, the probability of obtaining false positive results increases because the more tests you run, the greater the likelihood that some of them will generate significant results. To address this issue, multiple testing correction methods are employed to adjust the significance threshold for each test to control the overall false discovery rate (FDR).  This will help to reduce the chances of false positives and improves the accuracy of the analysis. With the **decideTests** function, we can identify genes that are differentially expressed while controlling the FDR. We set a threshold of 0.05 for adjusted p-values.

```{r}
results <- decideTests(efit, p.value = 0.05)
summary(results)
```
 
We observe that 7388 genes were upregulated and 4093 genes downregulated. 10530 genes were not significantly differentially expressed. The gene with the highest significance (p-value of 4.962176e-16), is gene '3423301'. Its corresponding log fold change value is -3.512177, indicating that the expression of this gene has significantly decreased in the tumor samples when compared to the control samples. The negative sign of the log fold change value confirms that the gene is downregulated, rather than upregulated.

## 6.2 MA-plots

```{r, name = "MA-plot plot of DE"}
par(mfrow = c(2, 1))
gridExtra::grid.arrange(limma::plotMA(efit), limma::plotMA(efit, ylim=c(-4,4)))
```
## 6.3 Volcano plots

Volcano plots are a data visualization tool to identify differentially expressed genes (DEGs). The plot displays the fold change in gene expression on the x-axis and the statistical significance of the change on the y-axis.Every gene is represented by a dot on the plot. The points that are of particular interest are those with a high |log fold change| and a significant level of statistical significance.

```{r, name = "Volcano plot of DE"}
par(mfrow = c(2,1))
gridExtra::grid.arrange volcanoplot(efit, xlim =c(-4,4), highlight = 25), nrow = 1)
```
## Conclusion of the plots

Upon examination, a surprisingly large number genes appear to have a high |log fold change| with a relatively high statistical significance in our dataset. This indicates that a considerable number of genes are experiencing significant up- and downregulation.

```{r, name = "Histogram and moderated plots of DE"}
hist(res_moderated_t$P.Value)

p6 <- ggplot(res_moderated_t, aes(x = AveExpr, y=logFC)) + geom_point() +
geom_point(data = subset(res_moderated_t, P.Value<0.05),
aes(x = AveExpr, y = logFC, color = "red")) +
geom_hline(yintercept = 0, colour = 'blue',
linetype = 'dashed', linewidth = 0.7)+
labs(title = 'Moderated') +
theme(plot.title = element_text(hjust = 0.5), legend.position = "none")
p8 <- ggplot(res_moderated_t, aes(x = logFC, y = -log10(P.Value))) +
geom_point()+
geom_point(data = subset(res_moderated_t, P.Value<0.05),
           aes(x = logFC, y = -log10(P.Value)), color = "red") +
geom_hline(yintercept = -log10(0.05))+
ylim(-0.5,7.5) +
labs(title = 'Moderated', x = 'log2 FC', y = '-log10(Pvalue)') +
theme(plot.title = element_text(hjust = 0.5), legend.position = "none")
par(mfrow = c(1,2))
gridExtra::grid.arrange(p6, p8, nrow = 1)
```

# 7. Gene set analysis (GSA)

Gene set analysis is performed to identify gene sets that are relevant as a group. We will check if a certain pathway is overrepresented among our set of interest, the differentially expressed genes, compared to the total set of genes. The GSA that we will perform, is an “overrepresentation analysis” (ORA).

We will use the Webgestalt tool to perform the gene set analysis. Since, there are a lot of differentially expressed genes (n = 11 481) we will opt to set the FDR cutoff to 0.05 and add an additional logFC cutoff of 4 to obtain an optimal set.

```{r}
# Select gene IDs for reference (all expressed genes)
gene_IDs_background <- Sign_Genes_ID$GeneID

# Select gene IDs for the set of interest
# Upregulated:
#Entrez_gene_ids_sign <- res$ENTREZID[(res$FDR < 0.01) & (res$logFC > 4)]
# Downregulated:
#Entrez_gene_ids_sign <- res$ENTREZID[(res$FDR < 0.01) & (res$logFC < -4)]
# General disruption:
gene_IDs_sign <- Sign_Genes_ID$GeneID[(Sign_Genes_ID$adj.P.Val < 0.05) & (abs(Sign_Genes_ID$logFC) > 2.5)]
length(gene_IDs_sign)

#remove all the non defined genes
gene_IDs_sign <- gene_IDs_sign[!is.na(gene_IDs_sign)]
#remove all the empty strings
gene_IDs_sign <- gene_IDs_sign[-which(gene_IDs_sign == "")]
gene_IDs_sign <- sort(as.character(gene_IDs_sign))

length(gene_IDs_sign)
```
By applying an FDR of 0.05 and a logFC of 2.5, 344 genes remain after the implementation of the extra quota, but only 166 are defined and are thus interesting for further analysis.

The gene IDs are written to .txt files which can be imported into the Webgestalt tool (select for Ensembl gene IDs) for GSA.

```{r}
# Write IDs to file
write.table(gene_IDs_background, file="./GSE68776/gene_IDs_background.txt", col.names=F, quote=F, row.names=F)
write.table(gene_IDs_sign, file="./GSE68776/gene_IDs_sign.txt",col.names=F, quote=F, row.names=F)
```

https://www.webgestalt.org/results/1683543507/

This final analysis shows us that the power of our GSA is not satisfactory since the FDR ratios of the pathways are above 0.05. Only the p53 signaling pathway still has a decent FDR of 0.086380 so this one is therefore interesting for interpretation. In the literature it is reported that genes involved in the p53 signaling pathway are mutated in approximately half of all human malignancies, including Ewing sarcoma. The p53 protein is a transcription factor known as the "guardian of the genome" because of its critical function in preserving genomic integrity and thus preventing tumor formation. The upregulation of this gene therefore could mean the formation of faster dividing cells, eventually leading to uncontrolled cell proliferation and cancer tumors. 
Several gene sets containing genes that are being upregulated had a much higher FDR, indicating a high chance of false positives. So the results should be interpreted with caution and may need further validation to confirm their significance. Various genes are reoccurring in the gene sets, including IGF1, TGFB2, ITGAV, CDK6 and AKT3. The NCBI summaries of these genes provide valuable insights into their potential roles in cancer. The expression of TGFB2 is expected, given that disruption of the TGF-beta/SMAD pathway has been linked to various human cancers. Similarly, ITGAV belongs to the integrin alpha chain family and has been implicated in regulating cancer progression. Also AKT3 and CDK6 are genes that regulate tumor formation, with AKT3 being activated by IGF1. 


# 8. Final conclusion of microarray analysis


We analyzed a raw dataset consisting of 21 samples (15 EWS & 6 control), and after quality control, we removed 2 samples, leaving us with 19 samples of decent quality. We then performed a needed normalization, especially because the controls differed a lot from the ES samples. At last, our differential expression analysis and gene set analysis showed that many genes were differentially expressed. This shows that gene regulation clearly differs in ES cells. However, due to lower power in GSA, we could not make definite statements about the connection between gene sets and Ewing sarcoma, except for maybe the p53 signaling pathway set, which generally deviates in cancer.



